### PR TITLE
add leaky_relu forward and backward in activation_op.cu

### PR DESCRIPTION
**Pull request by AnnaTrainingG**
### PR types

Performance optimization 
### PR changes

 OPs
### Describe

PR说明：本次添加leaky_relu正反向向量化实现，并新增了REGISTER_ACTIVATION_**GPU**_KERNEL宏用于注册向量化OP，相比于原始REGISTER_ACTIVATION\_**CUDA**_KERNEL 区别在于：ActivationGPUKernel（向量化实现）的调用；

leaky_relu新增向量化支持；
性能测试结果如下：

leaky backward | 单位us |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | --
数据类型 | pytorch | paddle old | paddle 1.8 | paddle new | (new-pytorch)/pytorch | (new-old)/old | (new-1.8/1.8
float32 | 2.918 | 2.581 | 2.385 | 2.244 | -23.10% | -13.06% | -5.91%
float32 | 113.29 | 117.43 | 116.37 | 114.33 | 0.92% | -2.64% | -1.75%
float32 | 2844.5 | 3142.6 | 3157.2 | 2864.7 | 0.71% | -8.84% | -9.26%
float32 | 287.74 | 298.92 | 296.15 | 289.8 | 0.72% | -3.05% | -2.14%
float16 | 2.236 | 6.173 | 2.193 | 1.823 | -18.47% | -70.47% | -16.87%
float16 | 58.494 | 67.084 | 65.767 | 58.669 | 0.30% | -12.54% | -10.79%
float16 | 1438.8 | 4696.2 | 2158.5 | 1463.7 | 1.73% | -68.83% | -32.19%
float16 | 146.34 | 165.74 | 164.52 | 148.73 | 1.63% | -10.26% | -9.60%
  |   |   |   |   |   |   |  
leaky forward | 单位us |   |   |   |   |   |  
数据类型 | pytorch | paddle old | paddle 1.8 | paddle new | (new-pytorch)/pytorch | (new-old)/old | (new-1.8/1.8
float32 | 2.846 | 2.49 | 1.824 | 1.834 | -35.56% | -26.35% | 0.55%
float32 | 77.749 | 78.571 | 81.747 | 77.263 | -0.63% | -1.66% | -5.49%
float32 | 1956.2 | 2103.8 | 2338.1 | 1944.9 | -0.58% | -7.55% | -16.82%
float32 | 197.6 | 201.29 | 210.73 | 196.41 | -0.60% | -2.42% | -6.80%
float16 | 2.186 | 5.964 | 1.814 | 1.691 | -22.64% | -71.65% | -6.78%
float16 | 47.231 | 48.692 | 47.756 | 41.399 | -12.35% | -14.98% | -13.31%
float16 | 1115.4 | 1352.2 | 1509.1 | 1024.7 | -8.13% | -24.22% | -32.10%
float16 | 116.08 | 120.26 | 120.52 | 104.09 | -10.33% | -13.45% | -13.63%

在进行float16计算的时候，使用half22float2转换成float计算，是为了精度考虑，此外使用float2计算并不会导致性能下降，性能测试结果如下：

forward | half2 | float2
-- | -- | --
[160800, 1, 7, 9] | 75.977us | 75.646us
[3, 513, 31, 31] | 12.975us | 12.469us
[8, 512, 157, 157] | 734.99us | 735.69us
[4, 128, 319, 31] | 38.928us | 39.265us
  |   |  
backward |   |  
[160800, 1, 7, 9] | 53.284us | 52.831us
[3, 513, 31, 31] | 8.9540us | 8.6850us
[8, 512, 157, 157] | 512.87us | 512.91us
[4, 128, 319, 31] | 26.944us | 27.039us

### CODE REVIEW

File: paddle/fluid/operators/activation_op.cu [hunk 1]
```diff
@@ -162,6 +166,113 @@ ReluGradGPUFunctor<float16>::Compute(const CudaVecType<float16>::type* out,
...
+ public:
+  LeakyReluGPUFunctor() { zero_ = static_cast<T>(0.0f); }
+
+  typename BaseActivationFunctor<T>::AttrPair GetAttrs() {
+    return {{"alpha", &alpha_}};
+  }
+  // leakyrelu forward : out = x > 0 ? x : x * alpha
+  __device__ __forceinline__ typename CudaVecType<T>::type Compute(
+      const typename CudaVecType<T>::type* x) {
+#ifdef __HIPCC__ || __CUDA_ARCH__ >= 350
```
Comment (zhangting2020) [inline paddle/fluid/operators/activation_op.cu]: relu有加`CUDA_VERSION >= 300`，这里是不是少写了？

Comment (AnnaTrainingG) [reply paddle/fluid/operators/activation_op.cu]: 重新修改了硬件判断

File: paddle/fluid/operators/activation_op.cu [hunk 2]
```diff
@@ -60,8 +64,13 @@ class ReluGPUFunctor : public BaseGPUFunctor<T> {
       const typename CudaVecType<T>::type* x);
 
   // when num % vecsize != 0 this func will be used
-  __device__ __forceinline__ T ComputeRemainder(const T x) {
-    return x > zero_ ? x : zero_;
+  __device__ __forceinline__ T ComputeRemainder(const T* x) {
```
Comment (Xreki) [inline paddle/fluid/operators/activation_op.cu]: 我觉得传值更简洁？数据加载的逻辑可以在公共的global函数里面处理，ComputeRemainder函数也能去掉`#ifdef`这个宏。

Comment (AnnaTrainingG) [reply paddle/fluid/operators/activation_op.cu]: 已经修改

File: paddle/fluid/operators/activation_op.cu [hunk 3]
```diff
@@ -124,44 +123,132 @@ class ReluGradGPUFunctor : public BaseGPUFunctor<T> {
...
+                     (in.y > zero_) ? (in.y) : (in.y) * alpha_,
+                     (in.z > zero_) ? (in.z) : (in.z) * alpha_,
+                     (in.w > zero_) ? (in.w) : (in.w) * alpha_);
+}
+
+template <>
+__device__ __forceinline__ CudaVecType<float16>::type
+LeakyReluGPUFunctor<float16>::Compute(const CudaVecType<float16>::type in) {
+  // leakyrelu forward : out = x > 0 ? x : x * alpha
+  const float2 xx = __half22float2(in);
```
Comment (Xreki) [inline paddle/fluid/operators/activation_op.cu:192]: 为什么要转成float来计算呢？

File: paddle/fluid/operators/activation_op.cu [hunk 4]
```diff
@@ -300,12 +421,28 @@ namespace plat = paddle::platform;
                                 ops::grad_functor<double>>,                 \
       ops::ActivationGradKernel<plat::CUDADeviceContext,                    \
                                 ops::grad_functor<plat::float16>>);
-
 FOR_EACH_ACTIVATION_OP(REGISTER_ACTIVATION_CUDA_KERNEL);
 
+#define REGISTER_ACTIVATION_GPU_KERNEL(act_type, op_name, functor,             \
```
Comment (Xreki) [inline paddle/fluid/operators/activation_op.cu:415]: 这2个宏的名字有点难以区分啊

**Review by Xreki — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/activation_op.cu (hunk 3 above)
Comment (AnnaTrainingG) [reply paddle/fluid/operators/activation_op.cu:192]: 为了保证计算的精度，另外转换成float计算并不会影响性能，测试结果如下：

forward | half2 | float2
-- | -- | --
[160800, 1, 7, 9] | 75.977us | 75.646us
[3, 513, 31, 31] | 12.975us | 12.469us
[8, 512, 157, 157] | 734.99us | 735.69us
[4, 128, 319, 31] | 38.928us | 39.265us
  |   |  
backward |   |  
[160800, 1, 7, 9] | 53.284us | 52.831us
[3, 513, 31, 31] | 8.9540us | 8.6850us
[8, 512, 157, 157] | 512.87us | 512.91us
[4, 128, 319, 31] | 26.944us | 27.039us

File: paddle/fluid/operators/activation_op.cu (hunk 4 above)
Comment (AnnaTrainingG) [reply paddle/fluid/operators/activation_op.cu:415]: 可以通过宏内调用的kernel进行区分ActivationGPUKernel为采用向量化的kernel实现，已经添加到PR说明

**Review by zhangting2020 — APPROVED** [review]
LGTM

**Pull request merged**